### DISCUSSION TITLE
RFC: Importing CSS from Third Party React Components

### DISCUSSION THREAD

**Discussion by Timer** [vercel/next.js#15208] [Ideas]
# RFC: Importing CSS from Third Party React Components

## Background

Next.js currently (as of July 2020) supports two methods of CSS imports:

1. **Global CSS**: Imports of Global CSS (e.g. CSS resets, themes) are restricted to an application's Custom `<App>` to avoid issues arising from ordering—importing Global CSS anywhere in your application would yield nondeterministic computed CSS styles based on the order you navigate through the application.
2. **CSS Modules**: Imports of "component-level" CSS (imported anywhere in the application) can be accomplished via CSS Modules. This prevents most nondeterminism issues, but doesn't completely eliminate them (best-effort). Its main purpose is to prevent a child-component from accidentally applying styles to a global selector like `div` (affecting the entire application).

## Use Case

Neither solution outlined above adequately solves for the use case of importing CSS from a third party component/library.

For example, using `@reach/slider` requires you import `@reach/slider/styles.css`.

Importing this file as Global CSS would increase the initial load size of every page in your application, even if the component is only used on a single page.

Importing this file as a CSS Module is not an option as the library would not understand having its classnames transformed into local versions.

## Proposal

Next.js should allow safely importing third-party/vendor CSS anywhere in the application.

Before we discuss a proposal, let's look at the current two ways CSS can be imported in a Next.js application:

```jsx
// This imports a file with global side effects, so it is only permitted in
// pages/_app.js.
// This allows selectors like "div { }" etc.
import "./css-reset.css";

// This imports a CSS Module file and exposes all of the class names in the
// `styles` object.
// Selectors like `.foo` in the CSS file must be referenced like `styles.foo`
// in first-party JavaScript.
import styles from "./Button.module.css";
```

For the `@reach/slider` use case, we'd want to be able to import `@reach/slider/styles.css` in an arbitrarily nested component file. This prevents it from being included on pages without a slider!

Due to the global CSS safeguards defined in **Background**, we cannot make the following "just work" in an arbitrarily nested component file:

```jsx
import "@reach/slider/styles.css";
```

Since 3rd-party CSS is out of your control, it must be scoped to prevent it from having global side effects in your application. Thus, we propose the following import style:

```jsx
import { className } from "@reach/slider/styles.css";

// or

import { className as sliderClassName } from "@reach/slider/styles.css";
```

This import style makes it **clear and explicit** that you're importing the 3rd party CSS with a framework-generated prefix, safely containing it from having side-effects in other parts of your application.

To use the vendor CSS, include the class name somewhere in the parent DOM:

```jsx
import { Slider } from "@reach/slider";
import { className } from "@reach/slider/styles.css";

function Example1() {
  return (
    <div className={className}>
      <Slider min={0} max={200} step={10} />
    </div>
  );
}

function Example2() {
  return <Slider className={className} min={0} max={200} step={10} />;
}
```

## Technical Details

This feature applies prefixing to the imported CSS selectors instead of transforming them into local, let's look at a code example to understand how it works:

**Input CSS**

```css
/* input.css */
div {
  background-color: blue;
}

.button {
  font-weight: bolder;
}
```

**Usage in JavaScript**

```jsx
import { className } from "./input.css";

className; // => 'input-XXX'
```

**Transformed Input (what's bundled)**

```css
/* input.css (transformed) */
.input-XXX div {
  background-color: blue;
}

.input-XXX .button {
  font-weight: bolder;
}
```

**Comment by arunoda** [comment]
@Timer this is a brilliant idea.

```js
import { Slider } from "@reach/slider";
import { className } from "@reach/slider/styles.css";

function Example1() {
  return (
    <div className={className}>
      <Slider min={0} max={200} step={10} />
    </div>
  );
}

function Example2() {
  return <Slider className={className} min={0} max={200} step={10} />;
}
```

I have no complaints or issues with approach. Let's go 🚀 .

**Reply by arunoda** [reply]
@Timer 

```js
function Example2() {
  return <Slider className={className} min={0} max={200} step={10} />;
}
```

Above only works, if only the component supports asking for a `className` right?

**Reply by sergiodxa** [reply]
I think that is the case when you wrap it in another element above the third-party component so you can load the styles.

**Reply by Janpot** [reply]
> Above only works, if only the component supports asking for a className right?

I think that would also require compiling to

```css
/* input.css (transformed) */
.input-XXX div, div.input-XXX {
  background-color: blue;
}

.input-XXX .button, .button.input-XXX {
  font-weight: bolder;
}
```
so that any class on the top level element is applied as well.

**Comment by giuseppeg** [comment]
The idea of adding an upper boundary is nice and would definitely work for 3rd party components that are leaf nodes eg. a Slider. However it might not work as nicely for components that accept bigger subtrees as `children` eg. a Modal dialog.

I think that this is a decent trade-off and probably [I wouldn't go further and overthink it](https://www.youtube.com/watch?v=tCVXp6gFD8o)! 

That said if you want a bullet-proof solution you might want explore the idea of using a custom pragma that would add a scoped class to every JSX element of the 3rd parties component. Something like this [https://codesandbox.io/s/k0ymm401zo](https://codesandbox.io/s/k0ymm401zo) but better engineered and prod ready. I think that this could be done with a webpack loader. The approach is very similar to styled-jsx's.

**Reply by giuseppeg** [reply]
Another case to consider is when components have side effects e.g. a modal component is open and adds a class to `body` to prevent scrolling (`overflow: hidden`):

```jsx
<body class="modal-open">
  <div className={className}>
    <Modal isOpen={true} />
  </div>
</body>
```

and

```css
/* modal.css (transformed) */
.modal-XXX .modal-open { overflow: hidden } /* this won't match any element */
```

In this case even my suggestion above would not work. With that I am not suggesting that you should fix all these issues, I am just posting them as they come to mind.

**Reply by pacocoursey** [reply]
Strongly agree that global selectors that will break with a className prefix. There needs to be an escape hatch to support this.

**Reply by justincy** [reply]
That sounds tough. You can't require 3rd party developers to denote which selectors are global; that'd be a non-starter. So Next.js apps would have to know which styles of their libs are global and have some way of informing Next; also tough but more realistic.

**Reply by giuseppeg** [reply]
I forgot to mention another issue with the OP proposal (I told Joe in private): this solution would break if components use portals which is a common scenario for modals tooltips etc.

Basically trying to scope styles can be a shoot in the foot. Next.js should probably just try to guarantee a deterministic insertion order of the styles. Maybe:

1. node_modules
2. _app.js (global)
3. .module.css

This can be tricky with code splitting and async components

**Comment by swyxio** [comment]
alternative syntax idea - `<style jsx src="@reach/slider/styles.css" />`

but i like that you are solving this!

**Reply by arso1er** [reply]
While I like that syntax, wouldn't this unecessarily increase bundle size for those not using styled-jsx?

**Reply by swyxio** [reply]
by how much are we talking? if it's under 3kb I'm fine w it. of course, I'd much prefer if styled-jsx had a static export option.

**Comment by paambaati** [comment]
@Timer Thanks for tackling this. One thing to consider here are complex components (like calendars/datepickers) that simply expect a CSS to be imported and don’t expose style props or class names as props. What would be the best way to solve for that?

**Comment by devongovett** [comment]
I like that you're thinking about this. I still feel like it's a lot to ask application authors to import the css for a library separately from the JS. That's an implementation detail of the library. Importing a component should come with the CSS automatically if it's imported by the library. Next.js is the only tool I know of where that's not true. Especially if it must to be done in the application root, it seems like it would be easy to get out of sync with what components you're actually using in your app (e.g. if you remove one from some page, you have to remember to also remove the CSS from the root?).

This is one of the first issues people ran into when trying out [react-spectrum](https://github.com/adobe/react-spectrum) in Next.js. In our case, we use CSS modules during development, but precompile them to global CSS to publish to npm. This works well with most tools, which treat CSS imports as global by default (e.g. CRA, Parcel, webpack with css-loader, etc.). I don't think I'd expect uncompiled CSS modules to work inside node_modules. If libraries publish global CSS that's not scoped somehow (e.g. by using CSS modules to build it, or something like BEM) then that's on them. 🤷 

Is there a technical reason why Next.js works differently from other tools, or just a philosophical one?

**Reply by pacocoursey** [reply]
>  I don't think I'd expect uncompiled CSS modules to work inside node_modules

That's not the goal of this RFC (though I guess it could be done?) Next.js should support importing any arbitrary CSS – uncompiled CSS modules could work if it's using vanilla features, or it could throw an error if using features like `composes`.

> Is there a technical reason why Next.js works differently from other tools

I don't have the technical insight to fully answer this, but the section in the background explains why Next.js ensures imported CSS is global in `_app`:

> Imports of Global CSS (e.g. CSS resets, themes) are restricted to an application's Custom `<App>` to avoid issues arising from ordering—importing Global CSS anywhere in your application would yield nondeterministic computed CSS styles based on the order you navigate through the application

The workaround for this non-determinism is ensuring all 3rd party CSS is "locally scoped" based on the file it's used in.

**Reply by devongovett** [reply]
I don't think it's realistic to force all third party libraries to support being used this way. There are many libraries already out there that import the CSS they need, and which already handle different orders they can be imported themselves. I would guess that most components that import CSS are designed such that the CSS is loaded together with the JS. And scoping your library CSS is just good practice. I don't think it should be up to Next.js to handle that. Library authors should be able to do so, and already need to do so for all other tools they support anyway. Seems like a great feature for application CSS, but for node_modules, I think it should be up to the library author.

**Reply by giuseppeg** [reply]
I agree with @devongovett on this

**Reply by arunoda** [reply]
>  If libraries publish global CSS that's not scoped somehow (e.g. by using CSS modules to build it, or something like BEM) then that's on them. 🤷

I disagree on this. When we are building an app, it's our responsibility whether we use our own stuff or third party modules. Here, Next.js is trying to help us with the safest approach.

This is what I'm talking about:

* Import a React component for a functionality in the dashboard
* It imports global CSS and it does not use CSS modules or similar
* Everything looks decent
* Ships the app
* Then the team ship more commits
* After 2 hours the sales team is worry. We are not getting any new sales
* After 10 hours of debugging, it found our that this global CSS import has a style which hides the credit card UI in the billing page

You might say, this should be covered with proper QA and UI testing setup. But frankly, I don't think everyone use such a sophisticated QA/ UI testing setup, which can detect each and every UI changes.

So, what Next.js is trying to do what any responsible framework should handle these issues. It's not a technical limitation or lazyness, but trying to be responsible.

**Reply by NMinhNguyen** [reply]
> There are many libraries already out there that import the CSS they need

I always thought that a library including `.css` imports in their code created a dependency on a bundler such as webpack with `css-loader` that would have to be applied to `node_modules`? Additionally, such a library probably wouldn't be easily consumable in native ESM environments or Node.js (SSR)? From what I've seen, most libraries that ship `.css` files to npm usually require the `.css` to be imported at application level (which is what `@reach/slider` does).

**Reply by devongovett** [reply]
The issue is that processing the CSS in some way that isn't expected by the library isn't going to work. If next.js changes the class names and the JS references them, then things are definitely not going to work. The only way to make that work is to pass the class names into each component, which would require complete restructuring of every library. If there's internal elements within a component, all of the classes for all sub-components would also need to be passed in somehow. This seems like a huge burden for users of any component library, and a huge amount of extra work for library authors to support next.js. It's just unrealistic to force library authors into the next.js way of doing things, especially when all other tools either work out of the box already (Create React App, Parcel, etc. support CSS imports in node_modules), or can be made to do so easily (Webpack via css-loader).

I understand the philosophical argument here but it's just not practical I'm afraid. Perhaps if Next.js doesn't want to allow this by default, there could be an option to enable it on a per module basis somewhere? This wouldn't be as good a DX for Next.js users but at least library authors would have somewhere to point their users about how to make them work.

**Reply by arunoda** [reply]
> This seems like a huge burden for users of any component library, and a huge amount of extra work for library authors to support next.js.

I think this is not the goal of this RFC. Component developers will write CSS for their components and Next.js is providing a safer way to include them in the user's app.

> If next.js changes the class names and the JS references them, then things are definitely not going to work.

I think you didn't fully understand the RFC. Next.js is not changing classnames. But it's creating a border for that CSS.

> Perhaps if Next.js doesn't want to allow this by default, there could be an option to enable it on a per module basis somewhere?

You can always customize Next.js config. But here we are trying to provide some good defaults.

**Reply by NMinhNguyen** [reply]
> The issue is that processing the CSS in some way that isn't expected by the library isn't going to work.

Although I agree with where you're coming from, there are (admittedly, somewhat rare and niche) cases where a library can't know how to inject styles, or rather where - when you're rendering into iframes, child windows spawned by the parent window, shadow DOM roots (e.g. if you export a Custom Element wrapping a React component) etc. These are the sorts of cases where decisions can only reliably be made at application level really, and it's something that insertion points enable (e.g. [1](https://material-ui.com/styles/advanced/#insertionpoint), [2](https://webpack.js.org/loaders/style-loader/#insert)).

**Reply by devongovett** [reply]
> I think this is not the goal of this RFC

The goal of the RFC isn't to make it as easy as possible to use third party libraries? IMO that should be the goal.

> Component developers will write CSS for their components and Next.js is providing a safer way to include them in the user's app.

I don't think many libraries will provide Next.js specific versions of their libraries. I would aim for compatibility with other tools here.

> I think you didn't fully understand the RFC. Next.js is not changing classnames.

Quote from the RFC: `This feature applies prefixing to the imported CSS selectors instead of transforming them into local`

> You can always customize Next.js config. But here we are trying to provide some good defaults.

Sure. I think people are struggling to customize it. And I'd argue that the proposal is *not* a good default because it's not compatible with other tools in the ecosystem, and requires a lot of additional work for both producers and consumers of libraries to support Next.js.

**Reply by devongovett** [reply]
> cases where a library can't know how to inject styles

The library shouldn't be injecting the styles into the DOM itself. Importing a CSS file is a declarative way to tell build tooling to handle that. Most tools handle this by injecting the CSS at the same time the JS loads. This could happen by injecting a `<style>` tag, but most of the time the CSS gets extracted and bundled together with other CSS that loads at the same time and loaded with a `<link rel="stylesheet">` element. None of this is up to the library though. The library just says "hey, I want this CSS", and the build tool for the app takes care of it.

**Reply by NMinhNguyen** [reply]
My point was more that, say you want to render a `@reach/slider` component into a new window (say, using React Portals) as well as the current window. If `@reach/slider` contained `import '@reach/slider/styles.css'` like `react-spectrum` does, then it would be impossible to specify the various targets (insertion points) given a single non-parameterised `import` that lives inside the library. What I'm trying to say is that sometimes this decision can't be baked into a library because it'd be fixed at build time whereas a user might have runtime requirements. And I believe this is why the RFC aims to enable Next.js users to perform CSS imports at application level (at the expense of an extra wrapping element).

**Reply by devongovett** [reply]
> render a @reach/slider component into a new window (say, using React Portals) as well as the current window

I don't see how the RFC as it stands addresses this. The CSS would load once, when the application imports it from its JS. It wouldn't load again when rendering through a portal into a new window without some serious magic from Next.js itself. And even if it did, why couldn't Next.js do the same for CSS imported from within a component? The import declaration tells you when the CSS is needed, so if you're going to implement magic like this I don't see why it couldn't apply everywhere.

**Reply by NMinhNguyen** [reply]
I haven't actually tried this yet because I typically use CSS in JS that gives me this flexibility out of the box, but something like an `import` (or maybe `require` to keep it synchronous) with webpack-specific loader syntax https://webpack.js.org/concepts/loaders/#inline would solve this use case? To clarify, you'd write a component that wraps `Slider` and handles importing the styles. Of course, the unfortunate thing here is you'd be relying on syntax that's tied to webpack, but at least it gives you an escape hatch in terms of configurability.

**Reply by justincy** [reply]
> > I think you didn't fully understand the RFC. Next.js is not changing classnames.
> 
> Quote from the RFC: `This feature applies prefixing to the imported CSS selectors instead of transforming them into local`

It would prepend a class to the existing selectors so that the CSS is no longer global but instead is effectively scoped to the component where it's used. A library with CSS that I want to use is Leaflet. It has a `.leaflet-popup` class. When imported, it might become `._23dcasd .leaflet-popup`. Leaflet JS can still create and find elements with the `.leaflet-popup` class but the CSS will only apply to locations where I insert the prefix that Next.js generated.

**Reply by pacocoursey** [reply]
> I don't think it's realistic to force all third party libraries to support being used this way
> I don't think many libraries will provide Next.js specific versions of their libraries

There's no "forcing" here. There's no motivation to change 3rd party libraries – this is a proposal for Next.js to support existing packages in a reasonable way.

> Quote from the RFC: `This feature applies prefixing to the imported CSS selectors instead of transforming them into local`

Sure, the classnames is literally being changed – but every CSS selector is prefixed equally.

>  The only way to make that work is to pass the class names into each component

I don't follow this. Compound components like `@reach/menu-button` don't include a CSS file for each component, they include 1 CSS file that bundles all of the necessary styles.

Usage should be like this:

```
import { Menu, MenuButton } from '@reach/menu-button'
import { className } from '@reach/menu-button/styles.css'
```

Not, as you suggest, like this:

```
import { Menu, MenuButton } from '@reach/menu-button'
import { MenuClassName, MenuButtonClassName } from '@reach/menu-button/styles.css'
```

**Reply by devongovett** [reply]
The forcing is in regards to component APIs. This RFC forces CSS and components to be distributed separately rather than together. The component itself does not import its own CSS, it must be able to be passed in. Lots of libraries aren't going to work with that out of the box.

**Comment by mattcarlotta** [comment]
100% agree with @devongovett points. On a related note, another RFC that was similar to this was requested [back in may](https://github.com/vercel/next.js/issues/13282) and there was some good? discussion within it. While this RFC attempts to expand imports to an agnostic `.css` file extension, I still hold the same sentiments as before.

## Summary

1.) Most third party libraries don't use local (module) css in production because the CSS is shared across different components. For example, Ant design's [Table](https://github.com/ant-design/ant-design/blob/master/components/table/style/index.tsx) uses styles from various other components (Button, Radio, Checkbox, ...etc) . Therefore, if a developer were to use a Table and a [Dropdown](https://github.com/ant-design/ant-design/blob/master/components/dropdown/style/index.tsx) component separately but within the same page, then they'll essentially be using duplicated CSS (one scoped to a Dropdown as a child to a Table and the other scoped to a Dropdown being its own parent). Which, unfortunately means larger production assets. 

2.) If the class names use dynamically generated `localIdentName`s, then it becomes harder for a developer to override 3rd party styles without using additional DOM markup (like using `id` to override a style). In addition, if a developer wishes to override these styles via a local/global stylesheet, then the import order (as of now) still isn't guaranteed and there still may be inconsistencies in the UI upon on page load.

3.) When it comes to component reusability, this RFC requires a different mental approach to importing components that most developers most likely won't be aware of. For example, the standard approach to import a 3rd party component is to (where CSS is imported by the package/a babel plugin or manually imported globally):

File 1:
```js
import { Table } from "antd";

...etc
<Table> ...</Table>
```

File 2:
```js
import { Table } from "antd";

...etc
<Table>...</Table>
```

While this RFC preaches...

File 1:
```js
import { Table } from "antd";
import { className } from "antd/table/styles.css";

const StyledTable = ({ children }) => (
  <Table className={className}>
    {children}
  </Table>
);

export default StyledTable;
```

File 2:
```js
import Table from "../components/Table";

...etc
<Table>...</Table>
```

File 3:
```js
import Table from "../components/Table";

...etc
<Table>...</Table>
```

While this reusability approach may seem incredibly arbitrary to some more experienced developers, you can bet most developers will not be aware of this Next limitation. It's not really a standard practice by other boilerplates (like the `create-react-app`) to require this import stylesheet granularity.

## Closing Thoughts

As I mentioned in the previous RFC, this has _very_ limited usage **as of now.** Most UI libraries expect their CSS to be globally imported. Therefore, this RFC may only cover a **rather small percentage** of libraries that support exporting CSS modules -- as for other libraries adopting this framework... eh... it's a big ask for a maintainer to refactor their entire library for a Next-specific limitation. **While I'm not against the feature's implementation, it just seems like another work-around to a bigger problem: Not being able to import/override 3rd party CSS reliably within Next**; and when it comes to wide-spread coverage/usage with what has already been published... it just seems too limited.

---

> You might say, this should be covered with proper QA and UI testing setup. But frankly, I don't think everyone use such a sophisticated QA/ UI testing setup, which can detect each and every UI changes.

> This is what I'm talking about:

> 1. Import a React component for a functionality in the dashboard
> 2. It imports global CSS and it does not use CSS modules or similar
> 3. Everything looks decent
> 4. Ships the app
> 5. Then the team ship more commits
> 6. After 2 hours the sales team is worry. We are not getting any new sales
> 7. After 10 hours of debugging, it found our that this global CSS import has a style which hides the credit card UI in the billing page

> So, what Next.js is trying to do what any responsible framework should handle these issues. It's not a technical limitation or lazyness, but trying to be responsible.

@arunoda That scenario should fall upon you (or your team), not upon Next. Next isn't responsible for your project's UI/UX. If some part of the UI doesn't function as **a result of your code, then that's on you.** I know there's a general push-back from most developers about QA, but if your project generates sales, then you darn-well better protect them by creating/running E2E tests to make sure your project functions as expected in production. I mean, come on! Look how well it's working out for Microsoft when they dumped their QA security patch team to save a few bucks: [Article](https://www.howtogeek.com/658194/windows-10s-new-update-is-deleting-peoples-files-again/), [Article](https://metro.co.uk/2020/03/13/microsoft-windows-10-update-causing-blue-screen-death-crashes-users-claim-12392206/)! Yikes, man! 😬

**Reply by justincy** [reply]
1) The issue of duplicate CSS is very valid. But the feature could easily be implemented in a smart way to avoid the duplication by sharing the same generated prefix across all usages in the app.

2) The className is a just a prefix. Styles can still be easily overridden globally. See https://github.com/vercel/next.js/discussions/15208#discussioncomment-41021

3) Perhaps I don't understand the point you're trying to make. It sounds like you're saying that a developer only needs to import 3rd party CSS once even if the 3rd party package is used multiple times. I believe this is incorrect. You would want to import the CSS every time you want it applied because the CSS isn't imported globally. If we can avoid duplication (as discussed in your first point) then this shouldn't be a problem.

**Reply by mattcarlotta** [reply]
> 1. The issue of duplicate CSS is very valid. But the feature could easily be implemented in a smart way to avoid the duplication by sharing the same generated prefix across all usages in the app.

I don't think this is a viable option, especially if the CSS files are separate. Next would have to compare every class name to any previous prefixed CSS and determine whether or not they're unique strings that deserve their own unique prefix or are shared strings that require a previously used prefix. This sounds like a caching nightmare and I'd presume would be very slow and ineffective.

> 3. Perhaps I don't understand the point you're trying to make. It sounds like you're saying that a developer only needs to import 3rd party CSS once even if the 3rd party package is used multiple times. I believe this is incorrect. You would want to import the CSS every time you want it applied because the CSS isn't imported globally. If we can avoid duplication (as discussed in your first point) then this shouldn't be a problem.

I think you misunderstood my 3rd point as what you've described is exactly what it's trying to illustrate: 

Even if the prefixes are the same, the CSS is still being applied locally (to a file). Therefore, if you want to reuse the component, then you would have to reimport the CSS. Meaning, if your index page uses a 3rd party `Button` and your about page uses the same third party button, then unless you're reimporting the CSS, the `Button` is unstyled. Something like the example below is an atypical requirement in current development practices (for 3rd party elements):

**pages/index.js**
```js
import { Button } from "antd";
import { classNames } from "antd/button";

<Button 
  className={classNames} 
  onClick={() => alert("Hi")}
>
  Hi
</Button>
```

**pages/about.js**
```js
import { Button } from "antd";
import { classNames } from "antd/button";

<Button 
  className={classNames} 
  onClick={() => alert("Bye")}
>
  Bye
</Button>
```
The work-around would be to do what I've mentioned in item 3 -- importing the third party item + CSS into a single project file and then importing that file instead of the third party package. Again, trivial, but not really a standard.

> 2. The className is a just a prefix. Styles can still be easily overridden globally. See [#15208 (reply in thread)](https://github.com/vercel/next.js/discussions/15208#discussioncomment-41021)

If I'm not mistaken, the idea is to keep the styles locally scoped to avoid global class name collisions and any potential side-effects:

> Its main purpose is to prevent a child-component from accidentally applying styles to a global selector like div (affecting the entire application).

In addition, globally styling a local class goes against the `pure` environment (guaranteed side-effect free) that Next is trying to establish.

**Comment by pacocoursey** [comment]
> safely containing it from having side-effects

This is my main concern - is this worth it? Is it Next.js' responsibility to do this? Right now Next.js does nothing to protect against these side-effects with styled-jsx:

```js
// side-effect-comp.js
const SideEffectComp = () => (
  <div>
    hi
    <style jsx global>{`body { background: red; }`}</style>
  </div>
)

// page.js
const Page = () => (
  <SideEffectComp />
)
```

Which is equivalent to importing unscoped CSS on a per-component basis:

```js
// side-effect-comp.js
import './side-effect-comp.css'

const SideEffectComp = () => (
  <div>
    hi
  </div>
)

// page.js
const Page = () => (
  <SideEffectComp />
)
```

As [Giuseppe pointed out](https://github.com/vercel/next.js/discussions/15208#discussioncomment-38472), prefixing with className potentially *breaks* 3rd party CSS, which seems like a non-starter, so there must be a way to support unscoped CSS in some way.

**Comment by justincy** [comment]
The RFC states: 

> Since 3rd-party CSS is out of your control, it must be scoped to prevent it from having global side effects in your application.

The previous [RFC for enhanced CSS support](https://github.com/vercel/next.js/issues/8626) stated something similar:

> Next.js will only allow you to import Global CSS within a custom pages/_app.js.
>
> This is a very important distinction (and a design flaw in other frameworks), as allowing Global CSS to be imported anywhere in your application means nothing could be code-split due to the cascading nature of CSS.
>
> This is an intentional constraint. Because when global CSS is imported in for example, a component, it will behave different when moving from one page to another.

I feel like this needs to be explained. Why does Next.js feel compelled to protect us from this? Are there situations where this is unavoidable? As a developer, it feels like dealing with this ought to be my job, not yours. I should be careful about what CSS I bring in. I should be capable of avoiding conflicts with global 3rd party CSS.

I'm not against this current proposal. It sounds pretty nice if we can get it working. I just want to understand how we got here.

Perhaps we can have both: scoped 3rd party CSS for most situations, unscoped 3rd party CSS for complex situations (like portals).

**Comment by dpatel4-lat** [comment]
How about supporting dynamic import in `pages/_app.js` and let all imports still be on `<App> ` but remove unused imports dynamically for each page request. Not sure if thats even possible.

**Comment by Timer** [comment]
See pull request here: https://github.com/vercel/next.js/pull/16993

**Reply by dipanshubhanot** [reply]
I am still running into some issues with the current implementation. I have a monorepo for multiple apps, so the node modules are accessed from a parent directory like ../../node_modules. In this case, I am still getting the Global CSS cannot be imported error message. Is there a way I can get around this?

**Comment by asader** [comment]
Hello, here is my workaround. I'm not sure if this will work for you, but it works for me.

```
const { getCssModuleLocalIdent } = require("next/dist/build/webpack/config/blocks/css/loaders/getCssModuleLocalIdent");

const regexEqual = (x, y) => {
  return (
    x instanceof RegExp &&
    y instanceof RegExp &&
    x.source === y.source &&
    x.global === y.global &&
    x.ignoreCase === y.ignoreCase &&
    x.multiline === y.multiline
  );
};

const nextConfig = {
  reactStrictMode: true,
  webpack: (config) => {
    const nextCssLoaders = config.module.rules.find((rule) => typeof rule.oneOf === 'object');

    if (nextCssLoaders) {
      const nextGlobalCssLoader = nextCssLoaders.oneOf.find(
        (rule) => rule.sideEffects === true && regexEqual(rule.test, /(?<!\.module)\.css$/)
      );
      if (nextGlobalCssLoader) {
        nextGlobalCssLoader.use.forEach((moduleLoader) => {
          if (
            moduleLoader.loader?.includes('css-loader') &&
            !moduleLoader.loader?.includes('postcss-loader')
          ) {
            moduleLoader.options.modules = moduleLoader.options.modules || {};
            moduleLoader.options.modules.getLocalIdent = getCssModuleLocalIdent;

          }
        });
      }
    }
    return config;
  }
};

/** @type {import('next').NextConfig} */
module.exports = nextConfig;
```
